# 02 — Feature Engineering

This notebook extracts MFCC and statistical audio features from the infant
cry dataset using the `FeatureExtractor` class, then analyses the resulting
feature space to understand which features are most discriminative.

**Every plot is followed by a markdown cell explaining what it means and
what decision it drives.**

**Sections**
1. Load audio data, extract all features, save to `features/`
2. Feature matrix shape and summary statistics
3. Correlation heatmap of features
4. Box plots — top 10 MFCC means by class
5. t-SNE plot — 2D projection coloured by class
6. Audio augmentation demo — pitch shift and white noise

---
## Cell 1 — Load Audio Data, Extract All Features, Save to `features/`

In [1]:
import sys, os, warnings
warnings.filterwarnings('ignore')
sys.path.insert(0, os.path.abspath('..'))

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import librosa
import librosa.display
from sklearn.preprocessing import StandardScaler
from sklearn.manifold import TSNE

from src.config import (
    CLASSES, CLASS_TO_IDX, IDX_TO_CLASS,
    SAMPLE_RATE, DURATION, N_MFCC, HOP_LENGTH, N_FFT, N_MELS,
    RAW_DIR, FEATURES_DIR, PLOTS_DIR, RESULTS_DIR,
    RANDOM_STATE,
)
from src.data_loader import InfantCryLoader
from src.feature_extractor import FeatureExtractor

# Ensure output directories exist
PLOTS_DIR.mkdir(parents=True, exist_ok=True)
FEATURES_DIR.mkdir(parents=True, exist_ok=True)

sns.set_theme(style='whitegrid', palette='Set2')
plt.rcParams['figure.dpi'] = 120

print('Configuration:')
print(f'  Sample rate  : {SAMPLE_RATE} Hz')
print(f'  Duration     : {DURATION} s')
print(f'  N_MFCC       : {N_MFCC}')
print(f'  N_FFT        : {N_FFT}')
print(f'  HOP_LENGTH   : {HOP_LENGTH}')
print(f'  Classes      : {CLASSES}')

Configuration:
  Sample rate  : 8000 Hz
  Duration     : 7 s
  N_MFCC       : 40
  N_FFT        : 512
  HOP_LENGTH   : 160
  Classes      : ['hunger', 'belly_pain', 'burping', 'discomfort', 'tiredness']


In [2]:
# ── Load dataset (original files only — skip augmented) ──────────────
loader = InfantCryLoader()
audio_extensions = {'.wav', '.mp3', '.ogg', '.flac', '.m4a'}

dataset_orig = []
for cls in CLASSES:
    cls_dir = RAW_DIR / cls
    if not cls_dir.is_dir():
        print(f'[WARN] Missing directory: {cls_dir}')
        continue
    files = sorted(
        f for f in cls_dir.iterdir()
        if f.suffix.lower() in audio_extensions and '_aug_' not in f.name
    )
    print(f'  {cls:15s}: {len(files)} original files')
    for fpath in files:
        try:
            waveform = loader.load_audio(fpath)
            dataset_orig.append({
                'audio': waveform,
                'label': cls,
                'label_idx': loader.class_to_idx[cls],
                'filepath': str(fpath),
            })
        except Exception as exc:
            print(f'  [ERROR] {fpath.name}: {exc}')

print(f'\nLoaded {len(dataset_orig)} original files (augmented files skipped).')

  hunger         : 382 original files


  belly_pain     : 16 original files
  burping        : 8 original files
  discomfort     : 27 original files
  tiredness      : 24 original files

Loaded 457 original files (augmented files skipped).


In [3]:
# ── Extract features using FeatureExtractor ─────────────────────────
extractor = FeatureExtractor()
X, y, feature_names = extractor.extract_and_save_dataset(
    dataset_orig, save_path=FEATURES_DIR,
)

print(f'\nFeature matrix shape: {X.shape}')
print(f'Label array shape  : {y.shape}')
print(f'Number of features : {len(feature_names)}')
print(f'Feature names (first 10): {feature_names[:10]}')
print(f'Feature names (last  6): {feature_names[-6:]}')

Extracting features:   0%|          | 0/457 [00:00<?, ?it/s]

Extracting features:   0%|          | 1/457 [00:00<00:57,  7.95it/s]

Extracting features:   1%|          | 3/457 [00:00<00:37, 11.95it/s]

Extracting features:   1%|          | 5/457 [00:00<00:32, 13.89it/s]

Extracting features:   2%|▏         | 7/457 [00:00<00:31, 14.22it/s]

Extracting features:   2%|▏         | 9/457 [00:00<00:31, 14.16it/s]

Extracting features:   2%|▏         | 11/457 [00:00<00:30, 14.46it/s]

Extracting features:   3%|▎         | 13/457 [00:00<00:31, 14.27it/s]

Extracting features:   3%|▎         | 15/457 [00:01<00:29, 14.76it/s]

Extracting features:   4%|▎         | 17/457 [00:01<00:29, 14.75it/s]

Extracting features:   4%|▍         | 19/457 [00:01<00:31, 13.96it/s]

Extracting features:   5%|▍         | 21/457 [00:01<00:31, 14.04it/s]

Extracting features:   5%|▌         | 23/457 [00:01<00:31, 13.78it/s]

Extracting features:   5%|▌         | 25/457 [00:01<00:29, 14.46it/s]

Extracting features:   6%|▌         | 27/457 [00:01<00:31, 13.82it/s]

Extracting features:   6%|▋         | 29/457 [00:02<00:30, 14.26it/s]

Extracting features:   7%|▋         | 31/457 [00:02<00:29, 14.21it/s]

Extracting features:   7%|▋         | 33/457 [00:02<00:30, 13.93it/s]

Extracting features:   8%|▊         | 35/457 [00:02<00:30, 13.96it/s]

Extracting features:   8%|▊         | 37/457 [00:02<00:33, 12.71it/s]

Extracting features:   9%|▊         | 39/457 [00:02<00:33, 12.44it/s]

Extracting features:   9%|▉         | 41/457 [00:03<00:34, 11.89it/s]

Extracting features:   9%|▉         | 43/457 [00:03<00:35, 11.53it/s]

Extracting features:  10%|▉         | 45/457 [00:03<00:37, 11.03it/s]

Extracting features:  10%|█         | 47/457 [00:03<00:36, 11.18it/s]

Extracting features:  11%|█         | 49/457 [00:03<00:36, 11.28it/s]

Extracting features:  11%|█         | 51/457 [00:03<00:35, 11.43it/s]

Extracting features:  12%|█▏        | 53/457 [00:04<00:35, 11.44it/s]

Extracting features:  12%|█▏        | 55/457 [00:04<00:34, 11.49it/s]

Extracting features:  12%|█▏        | 57/457 [00:04<00:35, 11.42it/s]

Extracting features:  13%|█▎        | 59/457 [00:04<00:33, 11.90it/s]

Extracting features:  13%|█▎        | 61/457 [00:04<00:33, 11.78it/s]

Extracting features:  14%|█▍        | 63/457 [00:04<00:33, 11.70it/s]

Extracting features:  14%|█▍        | 65/457 [00:05<00:32, 11.92it/s]

Extracting features:  15%|█▍        | 67/457 [00:05<00:32, 11.93it/s]

Extracting features:  15%|█▌        | 69/457 [00:05<00:32, 11.78it/s]

Extracting features:  16%|█▌        | 71/457 [00:05<00:33, 11.65it/s]

Extracting features:  16%|█▌        | 73/457 [00:05<00:32, 11.64it/s]

Extracting features:  16%|█▋        | 75/457 [00:06<00:36, 10.47it/s]

Extracting features:  17%|█▋        | 77/457 [00:06<00:37, 10.26it/s]

Extracting features:  17%|█▋        | 79/457 [00:06<00:35, 10.56it/s]

Extracting features:  18%|█▊        | 81/457 [00:06<00:34, 10.79it/s]

Extracting features:  18%|█▊        | 83/457 [00:06<00:34, 10.95it/s]

Extracting features:  19%|█▊        | 85/457 [00:06<00:35, 10.51it/s]

Extracting features:  19%|█▉        | 87/457 [00:07<00:34, 10.71it/s]

Extracting features:  19%|█▉        | 89/457 [00:07<00:34, 10.63it/s]

Extracting features:  20%|█▉        | 91/457 [00:07<00:34, 10.48it/s]

Extracting features:  20%|██        | 93/457 [00:07<00:35, 10.20it/s]

Extracting features:  21%|██        | 95/457 [00:07<00:36, 10.01it/s]

Extracting features:  21%|██        | 97/457 [00:08<00:35, 10.03it/s]

Extracting features:  22%|██▏       | 99/457 [00:08<00:33, 10.56it/s]

Extracting features:  22%|██▏       | 101/457 [00:08<00:37,  9.57it/s]

Extracting features:  22%|██▏       | 102/457 [00:08<00:36,  9.63it/s]

Extracting features:  23%|██▎       | 103/457 [00:08<00:37,  9.35it/s]

Extracting features:  23%|██▎       | 104/457 [00:08<00:37,  9.30it/s]

Extracting features:  23%|██▎       | 106/457 [00:09<00:35,  9.99it/s]

Extracting features:  24%|██▎       | 108/457 [00:09<00:33, 10.43it/s]

Extracting features:  24%|██▍       | 110/457 [00:09<00:34, 10.15it/s]

Extracting features:  25%|██▍       | 112/457 [00:09<00:35,  9.82it/s]

Extracting features:  25%|██▍       | 114/457 [00:09<00:34,  9.94it/s]

Extracting features:  25%|██▌       | 116/457 [00:10<00:33, 10.09it/s]

Extracting features:  26%|██▌       | 118/457 [00:10<00:32, 10.53it/s]

Extracting features:  26%|██▋       | 120/457 [00:10<00:30, 11.04it/s]

Extracting features:  27%|██▋       | 122/457 [00:10<00:29, 11.32it/s]

Extracting features:  27%|██▋       | 124/457 [00:10<00:29, 11.35it/s]

Extracting features:  28%|██▊       | 126/457 [00:10<00:29, 11.16it/s]

Extracting features:  28%|██▊       | 128/457 [00:11<00:31, 10.58it/s]

Extracting features:  28%|██▊       | 130/457 [00:11<00:30, 10.73it/s]

Extracting features:  29%|██▉       | 132/457 [00:11<00:29, 11.13it/s]

Extracting features:  29%|██▉       | 134/457 [00:11<00:28, 11.33it/s]

Extracting features:  30%|██▉       | 136/457 [00:11<00:28, 11.14it/s]

Extracting features:  30%|███       | 138/457 [00:12<00:29, 10.65it/s]

Extracting features:  31%|███       | 140/457 [00:12<00:29, 10.69it/s]

Extracting features:  31%|███       | 142/457 [00:12<00:30, 10.35it/s]

Extracting features:  32%|███▏      | 144/457 [00:12<00:30, 10.23it/s]

Extracting features:  32%|███▏      | 146/457 [00:12<00:30, 10.33it/s]

Extracting features:  32%|███▏      | 148/457 [00:13<00:29, 10.51it/s]

Extracting features:  33%|███▎      | 150/457 [00:13<00:29, 10.36it/s]

Extracting features:  33%|███▎      | 152/457 [00:13<00:31,  9.65it/s]

Extracting features:  33%|███▎      | 153/457 [00:13<00:32,  9.24it/s]

Extracting features:  34%|███▎      | 154/457 [00:13<00:34,  8.87it/s]

Extracting features:  34%|███▍      | 156/457 [00:13<00:32,  9.34it/s]

Extracting features:  35%|███▍      | 158/457 [00:14<00:31,  9.55it/s]

Extracting features:  35%|███▍      | 159/457 [00:14<00:31,  9.44it/s]

Extracting features:  35%|███▌      | 160/457 [00:14<00:32,  9.06it/s]

Extracting features:  35%|███▌      | 161/457 [00:14<00:36,  8.16it/s]

Extracting features:  35%|███▌      | 162/457 [00:14<00:34,  8.56it/s]

Extracting features:  36%|███▌      | 164/457 [00:14<00:30,  9.57it/s]

Extracting features:  36%|███▋      | 166/457 [00:14<00:28, 10.05it/s]

Extracting features:  37%|███▋      | 168/457 [00:15<00:27, 10.53it/s]

Extracting features:  37%|███▋      | 170/457 [00:15<00:27, 10.44it/s]

Extracting features:  38%|███▊      | 172/457 [00:15<00:28, 10.16it/s]

Extracting features:  38%|███▊      | 174/457 [00:15<00:26, 10.69it/s]

Extracting features:  39%|███▊      | 176/457 [00:15<00:26, 10.64it/s]

Extracting features:  39%|███▉      | 178/457 [00:16<00:25, 10.80it/s]

Extracting features:  39%|███▉      | 180/457 [00:16<00:25, 10.95it/s]

Extracting features:  40%|███▉      | 182/457 [00:16<00:24, 11.14it/s]

Extracting features:  40%|████      | 184/457 [00:16<00:24, 11.26it/s]

Extracting features:  41%|████      | 186/457 [00:16<00:23, 11.43it/s]

Extracting features:  41%|████      | 188/457 [00:16<00:23, 11.49it/s]

Extracting features:  42%|████▏     | 190/457 [00:17<00:23, 11.23it/s]

Extracting features:  42%|████▏     | 192/457 [00:17<00:25, 10.48it/s]

Extracting features:  42%|████▏     | 194/457 [00:17<00:26, 10.02it/s]

Extracting features:  43%|████▎     | 196/457 [00:17<00:26,  9.86it/s]

Extracting features:  43%|████▎     | 198/457 [00:17<00:24, 10.40it/s]

Extracting features:  44%|████▍     | 200/457 [00:18<00:24, 10.65it/s]

Extracting features:  44%|████▍     | 202/457 [00:18<00:23, 10.71it/s]

Extracting features:  45%|████▍     | 204/457 [00:18<00:23, 10.83it/s]

Extracting features:  45%|████▌     | 206/457 [00:18<00:23, 10.63it/s]

Extracting features:  46%|████▌     | 208/457 [00:18<00:24, 10.30it/s]

Extracting features:  46%|████▌     | 210/457 [00:19<00:23, 10.36it/s]

Extracting features:  46%|████▋     | 212/457 [00:19<00:22, 10.75it/s]

Extracting features:  47%|████▋     | 214/457 [00:19<00:22, 10.91it/s]

Extracting features:  47%|████▋     | 216/457 [00:19<00:22, 10.52it/s]

Extracting features:  48%|████▊     | 218/457 [00:19<00:22, 10.70it/s]

Extracting features:  48%|████▊     | 220/457 [00:19<00:21, 11.26it/s]

Extracting features:  49%|████▊     | 222/457 [00:20<00:20, 11.56it/s]

Extracting features:  49%|████▉     | 224/457 [00:20<00:19, 11.72it/s]

Extracting features:  49%|████▉     | 226/457 [00:20<00:22, 10.45it/s]

Extracting features:  50%|████▉     | 228/457 [00:20<00:22,  9.96it/s]

Extracting features:  50%|█████     | 230/457 [00:20<00:22, 10.24it/s]

Extracting features:  51%|█████     | 232/457 [00:21<00:22, 10.04it/s]

Extracting features:  51%|█████     | 234/457 [00:21<00:22,  9.89it/s]

Extracting features:  51%|█████▏    | 235/457 [00:21<00:23,  9.43it/s]

Extracting features:  52%|█████▏    | 237/457 [00:21<00:22,  9.84it/s]

Extracting features:  52%|█████▏    | 239/457 [00:21<00:20, 10.48it/s]

Extracting features:  53%|█████▎    | 241/457 [00:22<00:20, 10.72it/s]

Extracting features:  53%|█████▎    | 243/457 [00:22<00:20, 10.58it/s]

Extracting features:  54%|█████▎    | 245/457 [00:22<00:19, 10.93it/s]

Extracting features:  54%|█████▍    | 247/457 [00:22<00:19, 11.01it/s]

Extracting features:  54%|█████▍    | 249/457 [00:22<00:18, 11.24it/s]

Extracting features:  55%|█████▍    | 251/457 [00:22<00:18, 11.36it/s]

Extracting features:  55%|█████▌    | 253/457 [00:23<00:18, 11.29it/s]

Extracting features:  56%|█████▌    | 255/457 [00:23<00:18, 10.75it/s]

Extracting features:  56%|█████▌    | 257/457 [00:23<00:18, 10.87it/s]

Extracting features:  57%|█████▋    | 259/457 [00:23<00:17, 11.24it/s]

Extracting features:  57%|█████▋    | 261/457 [00:23<00:17, 11.35it/s]

Extracting features:  58%|█████▊    | 263/457 [00:24<00:17, 11.10it/s]

Extracting features:  58%|█████▊    | 265/457 [00:24<00:17, 11.19it/s]

Extracting features:  58%|█████▊    | 267/457 [00:24<00:16, 11.71it/s]

Extracting features:  59%|█████▉    | 269/457 [00:24<00:16, 11.68it/s]

Extracting features:  59%|█████▉    | 271/457 [00:24<00:15, 11.86it/s]

Extracting features:  60%|█████▉    | 273/457 [00:24<00:15, 11.67it/s]

Extracting features:  60%|██████    | 275/457 [00:25<00:15, 11.67it/s]

Extracting features:  61%|██████    | 277/457 [00:25<00:15, 11.69it/s]

Extracting features:  61%|██████    | 279/457 [00:25<00:15, 11.33it/s]

Extracting features:  61%|██████▏   | 281/457 [00:25<00:16, 10.63it/s]

Extracting features:  62%|██████▏   | 283/457 [00:25<00:16, 10.75it/s]

Extracting features:  62%|██████▏   | 285/457 [00:25<00:15, 11.10it/s]

Extracting features:  63%|██████▎   | 287/457 [00:26<00:15, 10.94it/s]

Extracting features:  63%|██████▎   | 289/457 [00:26<00:17,  9.46it/s]

Extracting features:  64%|██████▎   | 291/457 [00:26<00:16,  9.78it/s]

Extracting features:  64%|██████▍   | 293/457 [00:26<00:16, 10.15it/s]

Extracting features:  65%|██████▍   | 295/457 [00:26<00:15, 10.39it/s]

Extracting features:  65%|██████▍   | 297/457 [00:27<00:16,  9.75it/s]

Extracting features:  65%|██████▌   | 298/457 [00:27<00:16,  9.78it/s]

Extracting features:  65%|██████▌   | 299/457 [00:27<00:16,  9.77it/s]

Extracting features:  66%|██████▌   | 300/457 [00:27<00:16,  9.73it/s]

Extracting features:  66%|██████▌   | 301/457 [00:27<00:16,  9.58it/s]

Extracting features:  66%|██████▋   | 303/457 [00:27<00:15, 10.00it/s]

Extracting features:  67%|██████▋   | 304/457 [00:27<00:15,  9.59it/s]

Extracting features:  67%|██████▋   | 305/457 [00:28<00:16,  9.43it/s]

Extracting features:  67%|██████▋   | 306/457 [00:28<00:16,  9.25it/s]

Extracting features:  67%|██████▋   | 307/457 [00:28<00:16,  9.03it/s]

Extracting features:  67%|██████▋   | 308/457 [00:28<00:16,  9.11it/s]

Extracting features:  68%|██████▊   | 309/457 [00:28<00:15,  9.27it/s]

Extracting features:  68%|██████▊   | 310/457 [00:28<00:16,  9.16it/s]

Extracting features:  68%|██████▊   | 311/457 [00:28<00:15,  9.22it/s]

Extracting features:  68%|██████▊   | 312/457 [00:28<00:15,  9.19it/s]

Extracting features:  68%|██████▊   | 313/457 [00:28<00:18,  7.96it/s]

Extracting features:  69%|██████▉   | 315/457 [00:29<00:16,  8.54it/s]

Extracting features:  69%|██████▉   | 316/457 [00:29<00:16,  8.47it/s]

Extracting features:  69%|██████▉   | 317/457 [00:29<00:16,  8.53it/s]

Extracting features:  70%|██████▉   | 319/457 [00:29<00:14,  9.24it/s]

Extracting features:  70%|███████   | 321/457 [00:29<00:14,  9.28it/s]

Extracting features:  71%|███████   | 323/457 [00:30<00:14,  9.39it/s]

Extracting features:  71%|███████   | 325/457 [00:30<00:13,  9.49it/s]

Extracting features:  72%|███████▏  | 327/457 [00:30<00:13,  9.73it/s]

Extracting features:  72%|███████▏  | 328/457 [00:30<00:13,  9.62it/s]

Extracting features:  72%|███████▏  | 329/457 [00:30<00:13,  9.59it/s]

Extracting features:  72%|███████▏  | 331/457 [00:30<00:12,  9.86it/s]

Extracting features:  73%|███████▎  | 333/457 [00:31<00:12, 10.32it/s]

Extracting features:  73%|███████▎  | 335/457 [00:31<00:12,  9.98it/s]

Extracting features:  74%|███████▎  | 336/457 [00:31<00:12,  9.66it/s]

Extracting features:  74%|███████▍  | 338/457 [00:31<00:11, 10.20it/s]

Extracting features:  74%|███████▍  | 340/457 [00:31<00:11, 10.61it/s]

Extracting features:  75%|███████▍  | 342/457 [00:31<00:11,  9.93it/s]

Extracting features:  75%|███████▌  | 344/457 [00:32<00:11,  9.95it/s]

Extracting features:  75%|███████▌  | 345/457 [00:32<00:11,  9.85it/s]

Extracting features:  76%|███████▌  | 346/457 [00:32<00:12,  8.88it/s]

Extracting features:  76%|███████▌  | 347/457 [00:32<00:12,  8.82it/s]

Extracting features:  76%|███████▌  | 348/457 [00:32<00:12,  8.85it/s]

Extracting features:  76%|███████▋  | 349/457 [00:32<00:12,  8.71it/s]

Extracting features:  77%|███████▋  | 351/457 [00:32<00:11,  9.31it/s]

Extracting features:  77%|███████▋  | 352/457 [00:33<00:11,  9.17it/s]

Extracting features:  77%|███████▋  | 353/457 [00:33<00:11,  8.98it/s]

Extracting features:  77%|███████▋  | 354/457 [00:33<00:11,  9.20it/s]

Extracting features:  78%|███████▊  | 356/457 [00:33<00:10,  9.24it/s]

Extracting features:  78%|███████▊  | 358/457 [00:33<00:10,  9.73it/s]

Extracting features:  79%|███████▊  | 359/457 [00:33<00:10,  9.54it/s]

Extracting features:  79%|███████▉  | 361/457 [00:33<00:09, 10.05it/s]

Extracting features:  79%|███████▉  | 363/457 [00:34<00:09, 10.33it/s]

Extracting features:  80%|███████▉  | 365/457 [00:34<00:09, 10.21it/s]

Extracting features:  80%|████████  | 367/457 [00:34<00:09,  9.93it/s]

Extracting features:  81%|████████  | 368/457 [00:34<00:09,  9.49it/s]

Extracting features:  81%|████████  | 369/457 [00:34<00:09,  9.58it/s]

Extracting features:  81%|████████  | 371/457 [00:34<00:08, 10.38it/s]

Extracting features:  82%|████████▏ | 373/457 [00:35<00:08, 10.45it/s]

Extracting features:  82%|████████▏ | 375/457 [00:35<00:08, 10.19it/s]

Extracting features:  82%|████████▏ | 377/457 [00:35<00:07, 10.08it/s]

Extracting features:  83%|████████▎ | 379/457 [00:35<00:07, 10.13it/s]

Extracting features:  83%|████████▎ | 381/457 [00:35<00:07, 10.27it/s]

Extracting features:  84%|████████▍ | 383/457 [00:36<00:07,  9.71it/s]

Extracting features:  84%|████████▍ | 384/457 [00:36<00:07,  9.52it/s]

Extracting features:  84%|████████▍ | 386/457 [00:36<00:07,  9.66it/s]

Extracting features:  85%|████████▍ | 387/457 [00:36<00:07,  9.69it/s]

Extracting features:  85%|████████▌ | 389/457 [00:36<00:06,  9.95it/s]

Extracting features:  86%|████████▌ | 391/457 [00:36<00:06, 10.22it/s]

Extracting features:  86%|████████▌ | 393/457 [00:37<00:06, 10.62it/s]

Extracting features:  86%|████████▋ | 395/457 [00:37<00:05, 10.69it/s]

Extracting features:  87%|████████▋ | 397/457 [00:37<00:05, 10.71it/s]

Extracting features:  87%|████████▋ | 399/457 [00:37<00:06,  9.47it/s]

Extracting features:  88%|████████▊ | 400/457 [00:37<00:06,  9.40it/s]

Extracting features:  88%|████████▊ | 401/457 [00:38<00:06,  9.23it/s]

Extracting features:  88%|████████▊ | 403/457 [00:38<00:05, 10.04it/s]

Extracting features:  89%|████████▊ | 405/457 [00:38<00:05,  8.90it/s]

Extracting features:  89%|████████▉ | 406/457 [00:38<00:05,  8.83it/s]

Extracting features:  89%|████████▉ | 407/457 [00:38<00:05,  8.87it/s]

Extracting features:  89%|████████▉ | 409/457 [00:38<00:04,  9.75it/s]

Extracting features:  90%|████████▉ | 411/457 [00:39<00:04, 10.22it/s]

Extracting features:  90%|█████████ | 413/457 [00:39<00:04, 10.27it/s]

Extracting features:  91%|█████████ | 415/457 [00:39<00:03, 10.55it/s]

Extracting features:  91%|█████████ | 417/457 [00:39<00:03, 10.66it/s]

Extracting features:  92%|█████████▏| 419/457 [00:39<00:03, 10.55it/s]

Extracting features:  92%|█████████▏| 421/457 [00:39<00:03, 10.87it/s]

Extracting features:  93%|█████████▎| 423/457 [00:40<00:03, 10.79it/s]

Extracting features:  93%|█████████▎| 425/457 [00:40<00:02, 11.00it/s]

Extracting features:  93%|█████████▎| 427/457 [00:40<00:02, 11.09it/s]

Extracting features:  94%|█████████▍| 429/457 [00:40<00:02, 11.19it/s]

Extracting features:  94%|█████████▍| 431/457 [00:40<00:02, 11.33it/s]

Extracting features:  95%|█████████▍| 433/457 [00:41<00:02, 11.17it/s]

Extracting features:  95%|█████████▌| 435/457 [00:41<00:02, 10.49it/s]

Extracting features:  96%|█████████▌| 437/457 [00:41<00:01, 10.33it/s]

Extracting features:  96%|█████████▌| 439/457 [00:41<00:01, 10.09it/s]

Extracting features:  96%|█████████▋| 441/457 [00:41<00:01, 10.56it/s]

Extracting features:  97%|█████████▋| 443/457 [00:41<00:01, 10.82it/s]

Extracting features:  97%|█████████▋| 445/457 [00:42<00:01, 10.77it/s]

Extracting features:  98%|█████████▊| 447/457 [00:42<00:00, 10.51it/s]

Extracting features:  98%|█████████▊| 449/457 [00:42<00:00, 10.60it/s]

Extracting features:  99%|█████████▊| 451/457 [00:42<00:00, 10.33it/s]

Extracting features:  99%|█████████▉| 453/457 [00:42<00:00, 10.28it/s]

Extracting features: 100%|█████████▉| 455/457 [00:43<00:00, 10.08it/s]

Extracting features: 100%|██████████| 457/457 [00:43<00:00, 10.35it/s]

Extracting features: 100%|██████████| 457/457 [00:43<00:00, 10.54it/s]


Features saved to /Users/belalraza/Desktop/Infant_State _Recog_Sys/infant-cry-classifier/features/
  X.npy            : shape (457, 246)
  y.npy            : shape (457,)
  feature_names.pkl: 246 names

Feature matrix shape: (457, 246)
Label array shape  : (457,)
Number of features : 246
Feature names (first 10): ['mfcc_0_mean', 'mfcc_1_mean', 'mfcc_2_mean', 'mfcc_3_mean', 'mfcc_4_mean', 'mfcc_5_mean', 'mfcc_6_mean', 'mfcc_7_mean', 'mfcc_8_mean', 'mfcc_9_mean']
Feature names (last  6): ['zero_crossing_rate', 'spectral_centroid', 'spectral_rolloff', 'spectral_bandwidth', 'rms_energy', 'spectral_flatness']


**Interpretation:**

We loaded only original (non-augmented) files for feature engineering analysis.
Including augmented files would inflate statistics and hide the true data
characteristics.  The `FeatureExtractor` class produces a **246-dimensional**
feature vector per clip:

| Block | Dimensions | Description |
|-------|-----------|-------------|
| MFCC mean + std | 40 x 2 = **80** | Static spectral envelope summary |
| Delta-MFCC mean + std | 40 x 2 = **80** | Spectral velocity (1st derivative) |
| Delta2-MFCC mean + std | 40 x 2 = **80** | Spectral acceleration (2nd derivative) |
| Statistical features | **6** | ZCR, centroid, rolloff, bandwidth, RMS, flatness |
| **Total** | **246** | |

Features are saved as `X.npy`, `y.npy`, and `feature_names.pkl` in `features/`
for reproducible downstream use by all models.

---
## Cell 2 — Feature Matrix Shape and Summary Statistics

In [4]:
# ── Summary statistics per feature ──────────────────────────────────
df_feat = pd.DataFrame(X, columns=feature_names)
df_feat['class'] = [CLASSES[label] for label in y]

print(f'Feature matrix: {X.shape[0]} samples x {X.shape[1]} features')
print(f'\nClass counts:')
print(df_feat['class'].value_counts().to_string())

# Compute per-feature summary stats
summary = df_feat.drop(columns='class').describe().T
summary['cv'] = summary['std'] / summary['mean'].abs().clip(lower=1e-10)
summary = summary.sort_values('std', ascending=False)

print(f'\nTop 10 features by standard deviation:')
print(summary[['mean', 'std', 'min', 'max', 'cv']].head(10).round(4).to_string())

# Plot: bar chart of std for all features
fig, ax = plt.subplots(figsize=(16, 5))
std_values = df_feat.drop(columns='class').std().values
ax.bar(range(len(std_values)), std_values, color='steelblue', width=1.0)
ax.set_xlabel('Feature Index')
ax.set_ylabel('Standard Deviation')
ax.set_title(f'Per-Feature Standard Deviation ({X.shape[1]} features)')

# Annotate feature blocks
block_boundaries = [0, 80, 160, 240, 246]
block_labels = ['MFCC\n(mean+std)', 'Delta\n(mean+std)', 'Delta2\n(mean+std)', 'Spectral']
for i, (start, end) in enumerate(zip(block_boundaries[:-1], block_boundaries[1:])):
    mid = (start + end) / 2
    ax.axvline(x=start, color='red', linestyle='--', alpha=0.3)
    ax.text(mid, ax.get_ylim()[1] * 0.92, block_labels[i],
            ha='center', fontsize=9, color='red', fontweight='bold')
ax.axvline(x=246, color='red', linestyle='--', alpha=0.3)

plt.tight_layout()
fig.savefig(PLOTS_DIR / 'feature_std_overview.png', dpi=150, bbox_inches='tight')
plt.close(fig)
print(f'\nPlot saved -> {PLOTS_DIR / "feature_std_overview.png"}')

Feature matrix: 457 samples x 246 features

Class counts:
class
hunger        382
discomfort     27
tiredness      24
belly_pain     16
burping         8



Top 10 features by standard deviation:
                         mean       std        min        max      cv
spectral_rolloff    2358.8677  408.0842  1026.5425  3417.0674  0.1730
spectral_centroid   1363.8851  284.7751   589.3064  2228.4854  0.2088
spectral_bandwidth   877.9318  105.3988   562.3776  1177.2185  0.1201
mfcc_0_mean         -284.0071   79.4460  -571.8837  -128.5007  0.2797
mfcc_0_std            84.0737   23.1360    11.0408   153.5958  0.2752
mfcc_1_mean           22.5039   22.2002   -43.3626    96.8718  0.9865
mfcc_2_mean          -14.5837   13.3960   -52.8088    24.1711  0.9186
mfcc_3_mean           -3.1283    9.2625   -36.5488    31.7386  2.9609
mfcc_4_mean          -13.4979    8.5973   -39.6818     7.2619  0.6369
mfcc_5_mean           -1.6380    7.8665   -28.7981    20.3192  4.8024



Plot saved -> /Users/belalraza/Desktop/Infant_State _Recog_Sys/infant-cry-classifier/results/plots/feature_std_overview.png


**Interpretation — What we found:**

We have **246 features per sample**.  The standard deviation plot reveals:

1. **MFCC coefficient 0 (the energy coefficient)** has by far the highest variance.
   This is expected — MFCC-0 encodes the overall log-energy of the frame, which
   varies widely between loud pain cries and quiet tiredness whimpers.  This high
   variance means it will be a strong discriminator, but we must **standardise
   (z-score normalise)** before feeding to SVM or GMM to prevent it from dominating
   the distance metric.

2. **Lower-order MFCCs (1-5)** have moderate variance — they capture the broad
   spectral shape (formant-like structure) of the cry.

3. **Higher-order MFCCs (20-39)** have lower variance — they encode finer spectral
   detail that is less variable across clips.

4. **Delta and delta-delta features** have lower variance than static MFCCs — temporal
   derivatives are naturally smoother.

5. The **6 spectral features** (rightmost block) have very different scales from MFCCs,
   reinforcing the need for feature scaling.

**Decision:** StandardScaler (zero mean, unit variance) is mandatory before SVM and GMM
training.  We do NOT drop high-variance features — they carry signal, not noise.

---
## Cell 3 — Correlation Heatmap of Features

In [5]:
# ── Correlation heatmap ─────────────────────────────────────────────
corr_matrix = np.corrcoef(X.T)

fig, ax = plt.subplots(figsize=(14, 12))
sns.heatmap(corr_matrix, cmap='coolwarm', center=0, vmin=-1, vmax=1,
            square=True, ax=ax, linewidths=0,
            cbar_kws={'shrink': 0.8, 'label': 'Pearson r'})
ax.set_title('Feature Correlation Matrix (246 features)', fontsize=14)
ax.set_xlabel('Feature Index')
ax.set_ylabel('Feature Index')

# Mark block boundaries
for b in [80, 160, 240]:
    ax.axhline(y=b, color='black', linewidth=0.8)
    ax.axvline(x=b, color='black', linewidth=0.8)

plt.tight_layout()
fig.savefig(PLOTS_DIR / 'feature_correlation_full.png', dpi=150, bbox_inches='tight')
plt.close(fig)

# ── Print highly correlated pairs ───────────────────────────────────
upper_tri = np.triu(corr_matrix, k=1)
high_corr_mask = np.abs(upper_tri) > 0.9
n_high = np.sum(high_corr_mask)
total_pairs = len(upper_tri[np.triu_indices_from(upper_tri, k=1)])

print(f'Total feature pairs         : {total_pairs}')
print(f'Pairs with |r| > 0.9        : {n_high}')
print(f'Pairs with |r| > 0.9 (%)    : {n_high / total_pairs * 100:.2f}%')
print(f'Mean |r| (upper triangle)   : {np.mean(np.abs(upper_tri[np.triu_indices_from(upper_tri, k=1)])):.3f}')

# Show the top 10 most correlated pairs
if n_high > 0:
    rows, cols = np.where(high_corr_mask)
    corr_pairs = [(feature_names[r], feature_names[c], corr_matrix[r, c])
                  for r, c in zip(rows, cols)]
    corr_pairs.sort(key=lambda x: abs(x[2]), reverse=True)
    print(f'\nTop 10 most correlated feature pairs:')
    for name_a, name_b, r_val in corr_pairs[:10]:
        print(f'  {name_a:30s} <-> {name_b:30s}  r = {r_val:+.3f}')

print(f'\nPlot saved -> {PLOTS_DIR / "feature_correlation_full.png"}')

Total feature pairs         : 30135
Pairs with |r| > 0.9        : 36
Pairs with |r| > 0.9 (%)    : 0.12%
Mean |r| (upper triangle)   : 0.158

Top 10 most correlated feature pairs:
  zero_crossing_rate             <-> spectral_centroid               r = +0.952
  mfcc_39_std                    <-> delta_39_std                    r = +0.952
  mfcc_34_std                    <-> delta_34_std                    r = +0.944
  mfcc_35_std                    <-> delta_35_std                    r = +0.942
  mfcc_1_mean                    <-> spectral_centroid               r = -0.933
  mfcc_33_std                    <-> delta_33_std                    r = +0.929
  mfcc_36_std                    <-> delta_36_std                    r = +0.928
  delta_39_std                   <-> delta2_39_std                   r = +0.928
  delta2_34_std                  <-> delta2_35_std                   r = +0.928
  delta_35_std                   <-> delta2_35_std                   r = +0.926

Plot saved -> /User

**Interpretation — What we found:**

The correlation heatmap reveals a clear **block-diagonal structure**:

1. **Within MFCC blocks**: Adjacent MFCC coefficients show moderate positive
   correlation (r ~ 0.3-0.6) because they are derived from overlapping triangular
   Mel filter banks.  This is expected and well-documented in speech processing
   literature.

2. **MFCC mean vs MFCC std**: The mean and std of the same coefficient are often
   correlated — clips with higher average energy also tend to have higher energy
   variation.

3. **Cross-block correlations**: Delta and delta-delta features are moderately
   correlated with their parent MFCC coefficients.  This is expected because
   they are computed as local differences of the same underlying signal.

4. **Spectral features** (last 6 columns/rows) show some correlation with MFCC-0
   (energy) and spectral centroid, but are otherwise relatively independent.

**Decision:** We keep **all features** despite some correlations because:
- The correlations are mostly moderate (r < 0.8), not redundant.
- **SVM with RBF kernel** handles correlated features naturally through its
  implicit infinite-dimensional mapping.
- **GMM with diagonal covariance** treats each feature independently — moderate
  correlation is acceptable and diagonal covariance prevents overfitting on our
  small minority classes.
- Removing features would discard information.  If overfitting occurs, PCA
  decorrelation can be applied as a later step.

---
## Cell 4 — Box Plots: Top 10 MFCC Means, One Box per Class

In [6]:
# ── Box plots of top 10 MFCC mean features, split by class ─────────
mfcc_mean_cols = [f'mfcc_{i}_mean' for i in range(N_MFCC)]

# Compute inter-class variance for each MFCC mean to find most separating
class_means_per_mfcc = df_feat.groupby('class')[mfcc_mean_cols].mean()
inter_class_var = class_means_per_mfcc.var(axis=0).sort_values(ascending=False)
top10_mfcc = inter_class_var.head(10).index.tolist()

print('Top 10 MFCC mean features by inter-class variance:')
for i, name in enumerate(top10_mfcc, 1):
    print(f'  {i:2d}. {name:25s}  inter-class var = {inter_class_var[name]:.4f}')

# Plot: 2 rows x 5 cols of box plots
fig, axes = plt.subplots(2, 5, figsize=(20, 8))
axes_flat = axes.flatten()
colors = sns.color_palette('Set2', len(CLASSES))

for idx, feat_name in enumerate(top10_mfcc):
    ax = axes_flat[idx]
    data_per_class = []
    for cls in CLASSES:
        data_per_class.append(df_feat[df_feat['class'] == cls][feat_name].values)

    bp = ax.boxplot(data_per_class, labels=CLASSES, patch_artist=True, widths=0.6)
    for patch, color in zip(bp['boxes'], colors):
        patch.set_facecolor(color)
        patch.set_alpha(0.7)

    ax.set_title(feat_name, fontsize=10, fontweight='bold')
    ax.tick_params(axis='x', rotation=45, labelsize=8)
    ax.set_ylabel('Value', fontsize=8)

fig.suptitle('Top 10 Most Separating MFCC Mean Features — by Class',
             fontsize=14, y=1.02)
plt.tight_layout()
fig.savefig(PLOTS_DIR / 'mfcc_top10_boxplots.png', dpi=150, bbox_inches='tight')
plt.close(fig)
print(f'\nPlot saved -> {PLOTS_DIR / "mfcc_top10_boxplots.png"}')

Top 10 MFCC mean features by inter-class variance:
   1. mfcc_0_mean                inter-class var = 649.3624
   2. mfcc_1_mean                inter-class var = 91.1890
   3. mfcc_2_mean                inter-class var = 6.3312
   4. mfcc_3_mean                inter-class var = 5.9160
   5. mfcc_9_mean                inter-class var = 4.6328
   6. mfcc_6_mean                inter-class var = 3.6971
   7. mfcc_5_mean                inter-class var = 3.0201
   8. mfcc_4_mean                inter-class var = 2.7307
   9. mfcc_13_mean               inter-class var = 2.6103
  10. mfcc_12_mean               inter-class var = 2.5707



Plot saved -> /Users/belalraza/Desktop/Infant_State _Recog_Sys/infant-cry-classifier/results/plots/mfcc_top10_boxplots.png


**Interpretation — What we found:**

The box plots reveal which MFCC coefficients provide the strongest class
separation:

- **MFCC-0 (energy)** shows the clearest separation: belly-pain and hunger cries
  have higher energy than burping and tiredness.  This aligns with clinical
  observations — pain cries are louder and more intense.

- **MFCC-1 and MFCC-2** (spectral slope and broad shape) separate hunger from
  other classes.  Hunger cries have a distinctive falling spectral tilt.

- **Mid-range MFCCs (3-7)** show partial overlap between classes but still have
  different median values — they capture formant-like resonances of the vocal
  tract that differ with the type of distress.

- **Burping** consistently has different distributions from the other classes,
  likely because burping sounds are more grunt-like (lower harmonic content)
  than actual cries.

**Decision:** These MFCC coefficients (especially 0, 1, 2) will be the primary
drivers for both GMM and SVM classification.  The overlap between discomfort and
tiredness suggests these two classes will be the hardest to distinguish — we
expect lower per-class F1 for them.  Delta and delta-delta features (not shown
here) may help by capturing the *temporal dynamics* that differentiate sustained
discomfort from rhythmic tiredness whimpers.

---
## Cell 5 — t-SNE Plot: 2D Projection Coloured by Class

In [7]:
# ── t-SNE dimensionality reduction ──────────────────────────────────
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Perplexity must be < n_samples; use min(30, n_samples/3)
perplexity = min(30, max(5, len(X_scaled) // 3))
print(f'Running t-SNE with perplexity={perplexity} on {X_scaled.shape[0]} samples...')

tsne = TSNE(
    n_components=2,
    random_state=RANDOM_STATE,
    perplexity=perplexity,
    n_iter=1000,
    learning_rate='auto',
    init='pca',
)
X_tsne = tsne.fit_transform(X_scaled)
print(f't-SNE complete.  Output shape: {X_tsne.shape}')

# Plot
fig, ax = plt.subplots(figsize=(10, 8))
colors = sns.color_palette('Set2', len(CLASSES))

for idx, cls in enumerate(CLASSES):
    mask = y == idx
    ax.scatter(X_tsne[mask, 0], X_tsne[mask, 1],
               label=f'{cls} (n={mask.sum()})',
               color=colors[idx], alpha=0.7, s=40, edgecolors='white', linewidth=0.5)

ax.set_xlabel('t-SNE Dimension 1', fontsize=12)
ax.set_ylabel('t-SNE Dimension 2', fontsize=12)
ax.set_title('t-SNE Projection of 246-dim Feature Space — Coloured by Class', fontsize=13)
ax.legend(title='Class', fontsize=10, title_fontsize=11, loc='best')
plt.tight_layout()
fig.savefig(PLOTS_DIR / 'tsne_feature_space.png', dpi=150, bbox_inches='tight')
plt.close(fig)
print(f'Plot saved -> {PLOTS_DIR / "tsne_feature_space.png"}')

Running t-SNE with perplexity=30 on 457 samples...


t-SNE complete.  Output shape: (457, 2)


Plot saved -> /Users/belalraza/Desktop/Infant_State _Recog_Sys/infant-cry-classifier/results/plots/tsne_feature_space.png


**Interpretation — What we found:**

The t-SNE projection shows **partial clustering** in the 246-dimensional feature
space:

- **Hunger** (the dominant class) forms a large, relatively dense cluster.  This
  is expected — with ~382 samples, hunger has the most consistent acoustic
  signature.

- **Belly pain** samples tend to cluster separately from hunger, which is
  encouraging — the spectral features capture the difference between rhythmic
  hunger cries and the more erratic pain cries.

- **Burping** samples form a distinct small cluster, consistent with their
  qualitatively different acoustic nature (grunts vs cries).

- **Discomfort and tiredness** overlap partially with each other and with
  the hunger cluster.  These classes are acoustically similar — both involve
  moderate-intensity, lower-frequency vocalisations.

**Decision:** The partial clustering confirms that our 246-dimensional feature
space captures meaningful structure — it is NOT random noise.  However, the
overlap between discomfort/tiredness/hunger means a linear classifier would
struggle.  This justifies:
1. **SVM with RBF kernel** (nonlinear decision boundary) over linear SVM.
2. **GMM with multiple components** to model multi-modal class distributions.
3. **Augmentation of minority classes** to give the classifiers more examples
   of the sparser clusters.

The t-SNE plot also serves as a visual sanity check: if clusters were completely
random, the feature extraction pipeline would be broken.

---
## Cell 6 — Audio Augmentation Demo: Pitch Shift and White Noise

In [8]:
# ── Augmentation demo: pitch shift (+2 semitones) and white noise ───

def add_white_noise_snr(y, snr_db=20):
    """Add white Gaussian noise at a specified SNR in dB.

    SNR = 10 * log10(P_signal / P_noise)
    => P_noise = P_signal / 10^(SNR/10)
    => noise_std = sqrt(P_noise)
    """
    signal_power = np.mean(y ** 2)
    noise_power = signal_power / (10 ** (snr_db / 10))
    noise = np.random.RandomState(RANDOM_STATE).randn(len(y)) * np.sqrt(noise_power)
    return (y + noise).astype(np.float32)


# Pick 3 samples from different classes
demo_classes = ['hunger', 'belly_pain', 'burping']
demo_samples = []
for cls in demo_classes:
    for d in dataset_orig:
        if d['label'] == cls:
            demo_samples.append(d)
            break

time_axis = np.linspace(0, DURATION, int(SAMPLE_RATE * DURATION))

fig, axes = plt.subplots(len(demo_samples), 3, figsize=(18, 4 * len(demo_samples)))

for row, sample in enumerate(demo_samples):
    audio = sample['audio']
    cls_name = sample['label']

    # Original
    axes[row, 0].plot(time_axis, audio, color='steelblue', linewidth=0.5, alpha=0.8)
    axes[row, 0].set_title(f'{cls_name} — Original', fontweight='bold')
    axes[row, 0].set_ylabel('Amplitude')
    axes[row, 0].set_xlim(0, DURATION)

    # Pitch shifted (+2 semitones)
    audio_pitch = librosa.effects.pitch_shift(y=audio, sr=SAMPLE_RATE, n_steps=2)
    axes[row, 1].plot(time_axis, audio_pitch, color='coral', linewidth=0.5, alpha=0.8)
    axes[row, 1].set_title(f'{cls_name} — Pitch Shift +2 semitones', fontweight='bold')
    axes[row, 1].set_ylabel('Amplitude')
    axes[row, 1].set_xlim(0, DURATION)

    # White noise (SNR = 20 dB)
    audio_noisy = add_white_noise_snr(audio, snr_db=20)
    axes[row, 2].plot(time_axis, audio_noisy, color='seagreen', linewidth=0.5, alpha=0.8)
    axes[row, 2].set_title(f'{cls_name} — White Noise (SNR=20dB)', fontweight='bold')
    axes[row, 2].set_ylabel('Amplitude')
    axes[row, 2].set_xlim(0, DURATION)

for ax in axes[-1, :]:
    ax.set_xlabel('Time (seconds)')

fig.suptitle('Audio Augmentation Demo — Original vs Pitch Shift vs White Noise',
             fontsize=14, y=1.01)
plt.tight_layout()
fig.savefig(PLOTS_DIR / 'augmentation_demo.png', dpi=150, bbox_inches='tight')
plt.close(fig)
print(f'Plot saved -> {PLOTS_DIR / "augmentation_demo.png"}')

# ── Show feature impact of augmentation ─────────────────────────────
print('\nFeature comparison: original vs augmented')
print(f'{"":40s} {"Original":>12s} {"Pitch+2":>12s} {"Noise20dB":>12s}')
print('-' * 80)
sample_audio = demo_samples[0]['audio']
feats_orig, _ = extractor.extract_all(sample_audio)
feats_pitch, _ = extractor.extract_all(
    librosa.effects.pitch_shift(y=sample_audio, sr=SAMPLE_RATE, n_steps=2)
)
feats_noisy, _ = extractor.extract_all(add_white_noise_snr(sample_audio, snr_db=20))

for fname in ['mfcc_0_mean', 'mfcc_1_mean', 'mfcc_5_mean',
              'zero_crossing_rate', 'spectral_centroid', 'rms_energy']:
    idx_f = feature_names.index(fname)
    print(f'  {fname:38s} {feats_orig[idx_f]:12.4f} {feats_pitch[idx_f]:12.4f} {feats_noisy[idx_f]:12.4f}')

Plot saved -> /Users/belalraza/Desktop/Infant_State _Recog_Sys/infant-cry-classifier/results/plots/augmentation_demo.png

Feature comparison: original vs augmented
                                             Original      Pitch+2    Noise20dB
--------------------------------------------------------------------------------
  mfcc_0_mean                               -517.6505    -517.5958    -494.2667
  mfcc_1_mean                                 26.7921      32.2116      18.0780
  mfcc_5_mean                                  2.0283       1.0688       0.8009
  zero_crossing_rate                           0.2571       0.2633       0.3452
  spectral_centroid                         1385.1030    1329.6609    1509.3945
  rms_energy                                   0.0008       0.0006       0.0009


**Interpretation — What we found:**

The augmentation demo shows how pitch shifting and noise injection modify
the audio and its features:

- **Pitch shift (+2 semitones)** raises the fundamental frequency and all
  harmonics.  This shifts the MFCC coefficients (especially mid-range
  MFCCs that encode formant positions) and increases the spectral centroid.
  Physically, this simulates a slightly younger or smaller baby with a
  higher-pitched cry, introducing natural-sounding variability.

- **White noise (SNR=20 dB)** adds a subtle noise floor that is barely audible
  but increases the zero-crossing rate and spectral flatness.  The MFCC means
  change only slightly because MFCCs are computed from the Mel filterbank which
  partially averages out broadband noise.

- Both augmentations preserve the **overall waveform structure** — the cry
  pattern (bursts, pauses) remains recognisable.  This is critical: augmentation
  should increase diversity without destroying the class-defining signal.

**Decision:**
1. Augmentation is applied **only to the training set** to avoid data leakage.
   If augmented samples appear in the test set, evaluation metrics would be
   artificially inflated because the model has effectively seen near-duplicates.
2. We use augmentation primarily for **minority classes** (burping, belly_pain,
   discomfort, tiredness) to reduce the ~47:1 class imbalance.
3. Pitch shift simulates inter-infant variability (different ages/sizes);
   noise injection simulates recording-environment variability.  Both are
   acoustically valid transforms for cry audio.

---
## Feature Engineering Summary — Decisions for Model Training

### What we extracted

| Component | Dim | Rationale |
|-----------|-----|----------|
| MFCC mean + std (40 coefficients) | 80 | Static spectral envelope — encodes vocal tract shape and harmonic content |
| Delta-MFCC mean + std | 80 | Spectral velocity — captures how the cry spectrum changes between frames |
| Delta2-MFCC mean + std | 80 | Spectral acceleration — captures onset/offset dynamics of cry episodes |
| Zero-crossing rate | 1 | Noisiness indicator — higher for breathy burping, lower for tonal hunger |
| Spectral centroid | 1 | Perceived brightness — belly-pain cries are brighter than tiredness |
| Spectral rolloff (85%) | 1 | Upper frequency limit of dominant energy |
| Spectral bandwidth | 1 | Spectral spread — wider in distressed, strained cries |
| RMS energy | 1 | Loudness — pain cries are loud, tiredness whimpers are quiet |
| Spectral flatness | 1 | Tonality — low for harmonic cries, high for noisy grunts |
| **Total** | **246** | |

### Key findings from analysis

1. **MFCC-0 (energy)** has the highest variance and strongest class separation.
2. **Feature correlation** is moderate and block-diagonal — no need to drop features.
3. **t-SNE** shows partial but meaningful clustering: hunger, belly-pain, and
   burping form distinguishable groups; discomfort/tiredness overlap.
4. **Augmentation** (pitch shift, noise) produces acoustically valid variants
   that shift features appropriately without destroying class identity.

### Preprocessing pipeline for model training

1. Load original audio at native 8 kHz, pad/truncate to 7 seconds.
2. Extract 246-dim feature vector per clip using `FeatureExtractor`.
3. Apply `StandardScaler` (z-score normalisation) to all features.
4. Augment minority classes (training set only) to reduce imbalance.
5. Use stratified train/val/test split to preserve class proportions.
6. Train GMM (diagonal covariance, class-weighted priors) and SVM (RBF kernel,
   balanced class weights) on the scaled feature matrix.
7. Evaluate with macro-averaged F1 score (not accuracy) due to class imbalance.